Mathlib/Data/Finset/Sym.lean

```diff
@@ -0,0 +1,269 @@
+
+/-- Lifts a finset to `Sym α n`. `s.sym n` is the finset of all unordered tuples of cardinality `n`
+with elements in `s`. -/
+protected def sym (s : Finset α) : ∀ n, Finset (Sym α n)
+  | 0 => {∅}
+  | n + 1 => s.sup fun a ↦ Finset.image (Sym.cons a) (s.sym n)
+#align finset.sym Finset.sym
+
+@[simp]
+theorem sym_zero : s.sym 0 = {∅} := rfl
+#align finset.sym_zero Finset.sym_zero
+
+@[simp]
+theorem sym_succ : s.sym (n + 1) = s.sup fun a ↦ (s.sym n).image <| Sym.cons a := rfl
+#align finset.sym_succ Finset.sym_succ
+
+@[simp]
+theorem mem_sym_iff : m ∈ s.sym n ↔ ∀ a ∈ m, a ∈ s := by
+  induction' n with n ih
+  · refine' mem_singleton.trans ⟨_, fun _ ↦ Sym.eq_nil_of_card_zero _⟩
+    rintro rfl
+    exact fun a ha ↦ (Finset.not_mem_empty _ ha).elim
+  refine' mem_sup.trans ⟨_, fun h ↦ _⟩
+  · rintro ⟨a, ha, he⟩ b hb
+    rw [mem_image] at he
+    obtain ⟨m, he, rfl⟩ := he
+    rw [Sym.mem_cons] at hb
+    obtain rfl | hb := hb
+    · exact ha
+    · exact ih.1 he _ hb
+  · obtain ⟨a, m, rfl⟩ := m.exists_eq_cons_of_succ
+    exact
+      ⟨a, h _ <| Sym.mem_cons_self _ _,
+        mem_image_of_mem _ <| ih.2 fun b hb ↦ h _ <| Sym.mem_cons_of_mem hb⟩
+#align finset.mem_sym_iff Finset.mem_sym_iff
+
+@[simp]
+theorem sym_empty (n : ℕ) : (∅ : Finset α).sym (n + 1) = ∅ := rfl
+#align finset.sym_empty Finset.sym_empty
+
+theorem replicate_mem_sym (ha : a ∈ s) (n : ℕ) : Sym.replicate n a ∈ s.sym n :=
+  mem_sym_iff.2 fun b hb ↦ by rwa [(Sym.mem_replicate.1 hb).2]
+#align finset.replicate_mem_sym Finset.replicate_mem_sym
+
+protected theorem Nonempty.sym (h : s.Nonempty) (n : ℕ) : (s.sym n).Nonempty :=
+  let ⟨_a, ha⟩ := h
+  ⟨_, replicate_mem_sym ha n⟩
+#align finset.nonempty.sym Finset.Nonempty.sym
+
+@[simp]
+theorem sym_singleton (a : α) (n : ℕ) : ({a} : Finset α).sym n = {Sym.replicate n a} :=
+  eq_singleton_iff_unique_mem.2
+    ⟨replicate_mem_sym (mem_singleton.2 rfl) _, fun _s hs ↦
+      Sym.eq_replicate_iff.2 fun _b hb ↦ eq_of_mem_singleton <| mem_sym_iff.1 hs _ hb⟩
+#align finset.sym_singleton Finset.sym_singleton
+
+theorem eq_empty_of_sym_eq_empty (h : s.sym n = ∅) : s = ∅ := by
+  rw [← not_nonempty_iff_eq_empty] at h⊢
+  exact fun hs ↦ h (hs.sym _)
+#align finset.eq_empty_of_sym_eq_empty Finset.eq_empty_of_sym_eq_empty
+
+@[simp]
+theorem sym_eq_empty : s.sym n = ∅ ↔ n ≠ 0 ∧ s = ∅ := by
+  cases n
+  · exact iff_of_false (singleton_ne_empty _) fun h ↦ (h.1 rfl).elim
+  · refine ⟨fun h ↦ ⟨Nat.succ_ne_zero _, eq_empty_of_sym_eq_empty h⟩, ?_⟩
+    rintro ⟨_, rfl⟩
+    exact sym_empty _
+#align finset.sym_eq_empty Finset.sym_eq_empty
+
+@[simp]
+theorem sym_nonempty : (s.sym n).Nonempty ↔ n = 0 ∨ s.Nonempty := by
+  simp_rw [nonempty_iff_ne_empty, Ne.def]
+-- Porting note: using simp_rw does not work here, it does nothing...
+  rwa [sym_eq_empty, not_and_or, not_ne_iff]
+#align finset.sym_nonempty Finset.sym_nonempty
+
+-- Porting note: this is defined twice in the mathlib3 file, maybe this one is supposed to be
```

Comment:
Almost certainly; we should probably backport this

Reply:
https://github.com/leanprover-community/mathlib/pull/18426

Reply:
The mathlib3 PR fixing this is ready for review.

Reply:
Well, that was not the right fix. The obvious change to 
`alias sym_nonempty ↔ _ nonempty.sym`
tries to overwrite a lemma that in fact already exists. In mathlib3 that fails silently, but in mathlib4, that gives a blocking error during the #align. 

Probably the best solution is to remove the alias in both mathlib3 and mathlib4 since it is obviously doing nothing...

Reply:
@YaelDillies if I understand `git blame` correctly, you are the one that added the alias in the mathport origin file, see : https://github.com/leanprover-community/mathlib/blob/6516bcc54f718ddebd84af1882107268d237797a/src/data/finset/sym.lean#L63

As noted above, I think this `alias` is incorrect since it tries to overwrite a lemma that already exists and indeed it fails silently with Lean 3 (there is a similar one also a few lines below), but it blocks the port of this file to Lean 4 since #align fails because of it. 

If you agree, I would like to just delete the two aliases. In fact, they are not used anywhere and mathlib compiles fine without them, see https://github.com/leanprover-community/mathlib/pull/18603